### PR TITLE
add json schema for model config

### PR DESCRIPTION
**Pull request by caozhuozi**
fix: https://github.com/CloudNativeAI/model-spec/issues/46

### NOTE for Reviewers:

1. I was unable to find any auto generation involved from json schema to  Go structs or vice versa in image-spec.
2. The logic related to the schema in the `image-spec` repo primarily focuses on (example) validation and Go testing.

This PR attempts to align with that validation and testing logic from image-spec.

### Go files introduced

To mirror the structure of the `image-spec` repo, I’ve introduced the following Go files with some minor adjustments based on my own understanding:
| `model-spec`               | `image-spec`             | Role                                      |
|----------------------------|---------------------------|-------------------------------------------|
| `schema/schema.go`         | `schema/schema.go`        | Go metadata definition for the schema     |
| `schema/validator.go`      | `schema/validator.go`     | Go utility package to validate JSONs |
| `schema/config_test.go`    | `schema/config_test.go`   | Use `schema/validator.go`to test the config json schema with a variety of intentionally common mistakes and edge-cases config JSON examples       |
| `schema/example_test.go`   | `schema/spec_test.go`     | Markdown examples validation              |

**note**:

1. I renamed `schema/spec_test.go` (from image-spec) to `schema/example_test.go` in this repo. In image-spec, `spec_test.go` validates the examples embedded in Markdown files, but the name is a bit ambiguous. The renaming helps clarify the purpose of the file here.
2. I’ve taken most of the code directly from the image-spec and temporarily removed any unused code, such as those related to other specs like `layout`.

### NOTE for the JSON Schema file

For the JSON Schema definition in `config-schema.json`, I have embedded validation rules directly within the schema to make it **self-contained** and ensure consistent validation behavior across different languages.

This  differs from `image-spec`, where additional validation rules are implemented in Go code (`schema/validator.go`) rather than encoded in the JSON schema definition itself. 
For example, there is an environment variable regex validation in image-spec for the config. This validation is performed in Go code:: https://github.com/opencontainers/image-spec/blob/main/schema/validator.go#L180-L185
But this validation is **not** explicitly defined in the JSON Schema:
https://github.com/opencontainers/image-spec/blob/main/schema/config-schema.json#L41-L45
I don't think it is a good idea. Thats why I changed it in this repo.

### Additional Validation Rules Enforced in This Schema Besides `config.md`

I embedded the following validation rules directly into the JSON Schema:
- In `modelfs`:
  - `type` must be exactly `"layers"` (enforced via `"enum"`)
  - `diff_ids` must be a non-empty array (enforced via `"minItems": 1`)

### CODE REVIEW

**Comment by bergwolf** [comment]
Would you add a schema for all the spec? It makes sense to add them in one shot, better having a script to auto-generate the go files.

**Comment by caozhuozi** [comment]
HI, @bergwolf  Thank you for reviewing the PR!

> Would you add a schema for all the spec?

Are you referring to the manifest specification? Initially, I thought yes, but then I wondered given that the manifest spec itself follows the OCI manifest format, would it be redundant?

Hope I understand the question correctly. 🙇

**Comment by bergwolf** [comment]
I'm mostly considering auto-generating .go files from the json schema. Can we do it w/ the current schema?

**Comment by caozhuozi** [comment]
@bergwolf 
> I'm mostly considering auto-generating .go files from the json schema. Can we do it w/ the current schema?

I've checked out the image-spec JSON schema files, and it looks like they were added manually. There doesn't seem to be any  auto generation in both directions.

For example, see the commit history of config-schema.json: https://github.com/opencontainers/image-spec/commits/main/schema/config-schema.json

The only automation I've found related to the JSON schema files is a validation process. If you look at the [Makefile](https://github.com/opencontainers/image-spec/blob/main/Makefile#L76-L78) and the [spec_test.go](https://github.com/opencontainers/image-spec/blob/main/schema/spec_test.go), you can see it validates some example JSONs against the schema using the github.com/santhosh-tekuri/jsonschema/v5 package (see https://github.com/opencontainers/image-spec/blob/main/schema/validator.go#L27).

Of course, I'm not saying generating code wouldn't be a good idea. ;-)

**Comment by bergwolf** [comment]
@caozhuozi I see. Thanks for the details! Then, could you add the validation tests so that we make sure the `specs-go` aligns with the schema? It also ensures that we involve the spec and schema together in future changes.

**Comment by caozhuozi** [comment]
@bergwolf 
Sure. I will.

File: schema/config-schema.json [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            "type": "string"
+          },
+          "description": {
+            "type": "string"
+          }
+        },
+        "additionalProperties": false,
+        "required": [
+          "name",
+          "version"
```
Comment (bergwolf) [inline schema/config-schema.json]: Per `docs/config.md`, these are optional.

File: schema/config-schema.json [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        "$ref": "#/$defs/ModelFS"
+      },
+      "config": {
+        "$ref": "#/$defs/ModelConfig"
+      }
+    },
+    "additionalProperties": false,
+    "required": [
+      "descriptor",
+      "modelfs"
```
Comment (bergwolf) [inline schema/config-schema.json:21]: `config` is also required.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,7 @@
+.PHONY: validate-examples
+validate-examples: ## validate examples in the specification markdown files
+	go test ./schema/example_test.go
+
+.PHONY: test
+test:
+	go test ./...
```
Comment (gaius-qi) [inline Makefile]: image
A line or a block of lines is suggested for deletion.

File: schema/config-schema.json [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+          }
+        },
+        "additionalProperties": false,
+        "required": [
+          "type",
+          "diff_ids"
+        ]
+      }
+    }
+  }
```
Comment (gaius-qi) [inline schema/config-schema.json]: A line or a block of lines is suggested for deletion.

**Review by gaius-qi — CHANGES_REQUESTED** [review]

File: Makefile (hunk 3 above)
Comment (caozhuozi) [reply Makefile]: Its weird, I can't identify the trailing blank line on my end.

File: schema/config-schema.json (hunk 4 above)
Comment (caozhuozi) [reply schema/config-schema.json]: Its weird, I can't identify the trailing blank line on my end.

File: Makefile (hunk 3 above)
Comment (caozhuozi) [reply Makefile]: any suggestions on how to fix it? 
the following is my vs code editor screenshot. Im working on mac os.
image

Comment (caozhuozi) [reply Makefile]: following is the screenshot of`cat -e ./schema/config-schema.json` on my mac laptop:
image

Comment (bergwolf) [reply Makefile]: See the missing newline at the last line of the file

image

Comment (caozhuozi) [reply Makefile]: Thank you for helping me look into this small formatting issue!  I solved it by force appending a new line.
image

File: schema/config-schema.json (hunk 4 above)
Comment (caozhuozi) [reply schema/config-schema.json]: I just force append a new tail line, the issue has gone.

File: Makefile (hunk 3 above)
Comment (caozhuozi) [reply Makefile]: note: i just checked the repo of image-image, they also has a trailing blank new line:

**Review by gaius-qi — APPROVED** [review]
LGTM

**Review by bergwolf — APPROVED** [review]
lgtm, thanks!
